{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4c520538-3006-4502-a9a4-aa745672b937",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import sys\n",
    "import os\n",
    "import re\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import io"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33616cc4-8783-4c39-83bb-08b82bf33941",
   "metadata": {},
   "source": [
    "# SAMTOOLS COVERAGE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "84f3b8e6-0ae0-43a0-8e97-805f51c7fa64",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Reads</th>\n",
       "      <th>Bases Covered</th>\n",
       "      <th>Coverage (%)</th>\n",
       "      <th>Mean Depth</th>\n",
       "      <th>Mean Base Quality</th>\n",
       "      <th>Mean Mapping Quality</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Barcode</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Align Method</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>202555645</td>\n",
       "      <td>91754</td>\n",
       "      <td>234512</td>\n",
       "      <td>0.115777</td>\n",
       "      <td>0.122412</td>\n",
       "      <td>17.6</td>\n",
       "      <td>49.0</td>\n",
       "      <td>o1</td>\n",
       "      <td>b17</td>\n",
       "      <td>129S1_SvImJ</td>\n",
       "      <td>raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>188650360</td>\n",
       "      <td>100545</td>\n",
       "      <td>228749</td>\n",
       "      <td>0.121256</td>\n",
       "      <td>0.112919</td>\n",
       "      <td>18.7</td>\n",
       "      <td>46.6</td>\n",
       "      <td>o1</td>\n",
       "      <td>b17</td>\n",
       "      <td>129S1_SvImJ</td>\n",
       "      <td>raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>164074439</td>\n",
       "      <td>72576</td>\n",
       "      <td>192909</td>\n",
       "      <td>0.117574</td>\n",
       "      <td>0.099833</td>\n",
       "      <td>18.0</td>\n",
       "      <td>44.8</td>\n",
       "      <td>o1</td>\n",
       "      <td>b17</td>\n",
       "      <td>129S1_SvImJ</td>\n",
       "      <td>raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>160637212</td>\n",
       "      <td>62689</td>\n",
       "      <td>181534</td>\n",
       "      <td>0.113009</td>\n",
       "      <td>0.091981</td>\n",
       "      <td>18.7</td>\n",
       "      <td>47.7</td>\n",
       "      <td>o1</td>\n",
       "      <td>b17</td>\n",
       "      <td>129S1_SvImJ</td>\n",
       "      <td>raw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>157702868</td>\n",
       "      <td>101229</td>\n",
       "      <td>204053</td>\n",
       "      <td>0.129391</td>\n",
       "      <td>0.156047</td>\n",
       "      <td>17.6</td>\n",
       "      <td>47.5</td>\n",
       "      <td>o1</td>\n",
       "      <td>b17</td>\n",
       "      <td>129S1_SvImJ</td>\n",
       "      <td>raw</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome  Start        End   Reads  Bases Covered  Coverage (%)  \\\n",
       "0          1      1  202555645   91754         234512      0.115777   \n",
       "1          2      1  188650360  100545         228749      0.121256   \n",
       "2          3      1  164074439   72576         192909      0.117574   \n",
       "3          4      1  160637212   62689         181534      0.113009   \n",
       "4          5      1  157702868  101229         204053      0.129391   \n",
       "\n",
       "   Mean Depth  Mean Base Quality  Mean Mapping Quality Sample Barcode  \\\n",
       "0    0.122412               17.6                  49.0     o1     b17   \n",
       "1    0.112919               18.7                  46.6     o1     b17   \n",
       "2    0.099833               18.0                  44.8     o1     b17   \n",
       "3    0.091981               18.7                  47.7     o1     b17   \n",
       "4    0.156047               17.6                  47.5     o1     b17   \n",
       "\n",
       "     Reference Align Method  \n",
       "0  129S1_SvImJ          raw  \n",
       "1  129S1_SvImJ          raw  \n",
       "2  129S1_SvImJ          raw  \n",
       "3  129S1_SvImJ          raw  \n",
       "4  129S1_SvImJ          raw  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dpath = \"/scratch/indikar_root/indikar1/cstansbu/scpc_test/reports/coverage/\"\n",
    "\n",
    "names = {\n",
    "    '#rname' : 'Chromosome',\n",
    "    'startpos' : 'Start',\n",
    "    'endpos' : 'End',\n",
    "    'numreads' : 'Reads',\n",
    "    'covbases' : 'Bases Covered',\n",
    "    'coverage' : 'Coverage (%)',\n",
    "    'meandepth' : 'Mean Depth', \n",
    "    'meanbaseq' : 'Mean Base Quality',\n",
    "    'meanmapq' : 'Mean Mapping Quality',\n",
    "}\n",
    "\n",
    "df = []\n",
    "\n",
    "for f in os.listdir(dpath):\n",
    "    fpath = f\"{dpath}{f}\"\n",
    "    sample = f[:2]\n",
    "    cell = f[2:5]\n",
    "    ref = f.split(\".\")[1]\n",
    "    align_method = f.split(\".\")[2]\\\n",
    "\n",
    "    tmp = pd.read_csv(fpath, \n",
    "                      sep='\\t', \n",
    "                     )\n",
    "    tmp = tmp.rename(columns=names)\n",
    "    tmp['Sample'] = sample\n",
    "    tmp['Barcode'] = cell\n",
    "    tmp['Reference'] = ref\n",
    "    tmp['Align Method'] = align_method\n",
    "    df.append(tmp)\n",
    "\n",
    "df = pd.concat(df)\n",
    "\n",
    "# remove unplaced contigs\n",
    "chroms = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12',\n",
    "       '13', '14', '15', '16', '17', '18', '19', 'X']\n",
    "\n",
    "df = df[df['Chromosome'].isin(chroms)]\n",
    "\n",
    "pdf = pd.pivot_table(df, index=['Sample', \n",
    "                               'Barcode', \n",
    "                               'Reference', \n",
    "                               'Chromosome', \n",
    "                               'Start', \n",
    "                               'End'],\n",
    "                   columns='Align Method').reset_index()\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "24062fac-b99e-4458-9181-3d23173281ac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "saved: /scratch/indikar_root/indikar1/cstansbu/scpc_test/analysis/excel_reports/coverage_summary.xlsx\n"
     ]
    }
   ],
   "source": [
    "outpath = \"/scratch/indikar_root/indikar1/cstansbu/scpc_test/analysis/excel_reports/coverage_summary.xlsx\"\n",
    "with pd.ExcelWriter(outpath, engine='xlsxwriter') as writer:\n",
    "\n",
    "    # save each sample to a new tab\n",
    "    for sample_names, group in pdf.groupby(['Sample', 'Reference']):\n",
    "        sample, ref = sample_names\n",
    "        group.to_excel(writer, sheet_name=f\"Sample_{sample}_ref_{ref}\", index=True)\n",
    "\n",
    "print(f'saved: {outpath}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "47db370d-c6ed-4322-abff-a041602fc3fd",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "'break' outside loop (668683560.py, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn[4], line 1\u001b[0;36m\u001b[0m\n\u001b[0;31m    break\u001b[0m\n\u001b[0m    ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m 'break' outside loop\n"
     ]
    }
   ],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf9b0fa5-e4f2-4a44-bbd9-8315b92d52fd",
   "metadata": {},
   "source": [
    "# Histograms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3e9de18-693a-45cb-b652-9d8807388bae",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dbd54c9-d5d0-4855-aecc-5270370c047b",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_cols = [\n",
    "    'Bases Covered', \n",
    "    'Coverage (%)',\n",
    "    'Mean Depth', \n",
    "    'Mean Base Quality', \n",
    "    'Mean Mapping Quality', \n",
    "]\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 200\n",
    "plt.rcParams['figure.figsize'] = 3.5, 2.5\n",
    "\n",
    "for c in plot_cols:\n",
    "    sns.histplot(data=df, \n",
    "                 x=c,\n",
    "                 bins=31,\n",
    "                 log_scale=True)\n",
    "\n",
    "    plt.title(c)\n",
    "    sns.despine()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "528b44f6-8f9b-4107-9f25-4536ecd092d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_cols = [\n",
    "    'Bases Covered', \n",
    "    'Coverage (%)',\n",
    "    'Mean Depth', \n",
    "    'Mean Base Quality', \n",
    "    'Mean Mapping Quality', \n",
    "]\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 200\n",
    "plt.rcParams['figure.figsize'] = 3.5, 2.5\n",
    "\n",
    "for c in plot_cols:\n",
    "    sns.histplot(data=df, \n",
    "                 x=c,\n",
    "                 bins=31,\n",
    "                 hue='Align Method',\n",
    "                 log_scale=True,)\n",
    "                 # kde=True)\n",
    "    \n",
    "    plt.title(c)\n",
    "    sns.despine()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d07671ec-aee8-45b1-8ca6-06c2562eaa61",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_cols = [\n",
    "    'Bases Covered', \n",
    "    'Coverage (%)',\n",
    "    'Mean Depth', \n",
    "    'Mean Base Quality', \n",
    "    'Mean Mapping Quality', \n",
    "]\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 200\n",
    "plt.rcParams['figure.figsize'] = 3.5, 2.5\n",
    "\n",
    "for c in plot_cols:\n",
    "    sns.histplot(data=df, \n",
    "                 x=c,\n",
    "                 bins=31,\n",
    "                 hue='Sample',\n",
    "                 log_scale=True,)\n",
    "\n",
    "    plt.title(c)\n",
    "    sns.despine()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd141714-4059-436a-8627-7d14ffcc4e66",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_cols = [\n",
    "    'Bases Covered', \n",
    "    'Coverage (%)',\n",
    "    'Mean Depth', \n",
    "    'Mean Base Quality', \n",
    "    'Mean Mapping Quality', \n",
    "]\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 200\n",
    "plt.rcParams['figure.figsize'] = 3.5, 2.5\n",
    "\n",
    "for c in plot_cols:\n",
    "    sns.histplot(data=df, \n",
    "                 x=c,\n",
    "                 bins=31,\n",
    "                 hue='Reference',\n",
    "                 log_scale=True)\n",
    "\n",
    "    plt.title(c)\n",
    "    sns.despine()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04b9194c-4a31-454f-b826-d5162f3d6d0d",
   "metadata": {},
   "source": [
    "# by chrom"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e1a1d84-f901-4401-b3d0-e37e2cec3f81",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_cols = [\n",
    "    'Bases Covered', \n",
    "    'Coverage (%)',\n",
    "    'Mean Depth', \n",
    "    'Mean Base Quality', \n",
    "    'Mean Mapping Quality', \n",
    "]\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 200\n",
    "plt.rcParams['figure.figsize'] = 7, 2.5\n",
    "\n",
    "for c in plot_cols:\n",
    "    sns.boxplot(data=df,\n",
    "                x='Chromosome',\n",
    "                y=c,\n",
    "                hue='Align Method',\n",
    "                showfliers=False,)\n",
    "\n",
    "    plt.title(c)\n",
    "    sns.despine()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4ca3b59-da6e-41f3-9d7e-17dd310c78bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_cols = [\n",
    "    'Bases Covered', \n",
    "    'Coverage (%)',\n",
    "    'Mean Depth', \n",
    "    'Mean Base Quality', \n",
    "    'Mean Mapping Quality', \n",
    "]\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 200\n",
    "plt.rcParams['figure.figsize'] = 7, 2.5\n",
    "\n",
    "for c in plot_cols:\n",
    "    sns.boxplot(data=df,\n",
    "                x='Chromosome',\n",
    "                y=c,\n",
    "                hue='Reference',\n",
    "                showfliers=False,)\n",
    "\n",
    "    plt.title(c)\n",
    "    sns.despine()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b4569d7-235f-4b72-9a8a-ecfb94278e3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_cols = [\n",
    "    'Bases Covered', \n",
    "    'Coverage (%)',\n",
    "    'Mean Depth', \n",
    "    'Mean Base Quality', \n",
    "    'Mean Mapping Quality', \n",
    "]\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 200\n",
    "plt.rcParams['figure.figsize'] = 7, 2.5\n",
    "\n",
    "for c in plot_cols:\n",
    "    sns.boxplot(data=df,\n",
    "                x='Chromosome',\n",
    "                y=c,\n",
    "                hue='Sample',\n",
    "                showfliers=False,)\n",
    "\n",
    "    plt.title(c)\n",
    "    sns.despine()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a1bc8a0-488c-4fec-9e65-94d1287d7141",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
