{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "791631da-48fb-4a84-a99d-91c524412c04",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import re\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pysam\n",
    "from Bio.Seq import Seq\n",
    "from Bio import Restriction\n",
    "from Bio import SeqIO\n",
    "from rapidfuzz import fuzz\n",
    "import rapidfuzz\n",
    "import regex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "80f26c64-7712-4d65-b928-95f7748b727d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BP MISMATCHES: 1 SCORE: 97.9\n",
      "BP MISMATCHES: 2 SCORE: 95.7\n",
      "BP MISMATCHES: 3 SCORE: 93.3\n",
      "BP MISMATCHES: 4 SCORE: 90.9\n",
      "BP MISMATCHES: 5 SCORE: 88.4\n",
      "BP MISMATCHES: 6 SCORE: 85.7\n",
      "BP MISMATCHES: 7 SCORE: 82.9\n",
      "BP MISMATCHES: 8 SCORE: 80.0\n",
      "BP MISMATCHES: 9 SCORE: 76.9\n",
      "BP MISMATCHES: 10 SCORE: 73.7\n"
     ]
    }
   ],
   "source": [
    "nbp = 24\n",
    "n_mismatch = 10\n",
    "ref_barcode = 'A' * nbp\n",
    "test_barcode = 'A' * nbp\n",
    "\n",
    "for i in range(n_mismatch):\n",
    "    test_barcode = test_barcode[:i] + 'T' + test_barcode[i+1:]\n",
    "    result = fuzz.partial_ratio_alignment(test_barcode, ref_barcode)\n",
    "    print(f\"BP MISMATCHES: {i+1} SCORE: {result.score:.1f}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9b4f20f6-91a7-4663-86b5-ba4086e2c454",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ScoreAlignment(score=100.0, src_start=0, src_end=24, dest_start=0, dest_end=24)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "barcode = \"AAGAAAGTTGTCGGTGTCTTTGTG\"\n",
    "read_seq = \"AAGAAAGTTGTCGGTGTCTTTGTGTGCTGAAGAAGTTGTCGGTGTCTTTGTGTTAACCTTTCTGTTGGTGCTGATATTCCGGTGCTGAAGATAGAGCGACAGGCAAGTAGGTTAACACAAAGACGCCGACAACTTTCTTCAGAAGAAAGTTGTCGGTGTCTTTGTG\"\n",
    "\n",
    "result = fuzz.partial_ratio_alignment(read_seq, barcode)\n",
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5b95743f-34d1-4bb2-9828-792a25d3ee3c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_name</th>\n",
       "      <th>read_seq_length</th>\n",
       "      <th>enzyme_matches</th>\n",
       "      <th>enzyme_pos</th>\n",
       "      <th>total_bc_matches</th>\n",
       "      <th>forward_bc_pos</th>\n",
       "      <th>forward_bc_score</th>\n",
       "      <th>forward_bc_macthes</th>\n",
       "      <th>reverse_bc_pos</th>\n",
       "      <th>reverse_bc_score</th>\n",
       "      <th>reverse_bc_macthes</th>\n",
       "      <th>mean_base_quality</th>\n",
       "      <th>median_base_quality</th>\n",
       "      <th>min_base_quality</th>\n",
       "      <th>max_base_quality</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6ff9aeed-b85d-415a-aa59-3ff86bbe727d</td>\n",
       "      <td>532</td>\n",
       "      <td>1</td>\n",
       "      <td>492</td>\n",
       "      <td>8</td>\n",
       "      <td>34;93;152;211;270</td>\n",
       "      <td>100.0;100.0;100.0;100.0;100.0</td>\n",
       "      <td>5</td>\n",
       "      <td>343;402;458</td>\n",
       "      <td>100.0;100.0;100.0</td>\n",
       "      <td>3</td>\n",
       "      <td>30</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5faf5728-3f2e-4d37-8f62-574673a1662e</td>\n",
       "      <td>286</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>4</td>\n",
       "      <td>35</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1</td>\n",
       "      <td>108;167;226</td>\n",
       "      <td>100.0;100.0;100.0</td>\n",
       "      <td>3</td>\n",
       "      <td>38</td>\n",
       "      <td>42</td>\n",
       "      <td>2</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4e3a9167-dac1-429e-9351-ce55485a4fae</td>\n",
       "      <td>416</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>5</td>\n",
       "      <td>92;151;210</td>\n",
       "      <td>100.0;100.0;100.0</td>\n",
       "      <td>3</td>\n",
       "      <td>283;342</td>\n",
       "      <td>100.0;100.0</td>\n",
       "      <td>2</td>\n",
       "      <td>36</td>\n",
       "      <td>41</td>\n",
       "      <td>4</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>f536f761-c128-4cde-a247-8b3a924b3399</td>\n",
       "      <td>261</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>3</td>\n",
       "      <td>38;96</td>\n",
       "      <td>100.0;97.87</td>\n",
       "      <td>2</td>\n",
       "      <td>183</td>\n",
       "      <td>95.83</td>\n",
       "      <td>1</td>\n",
       "      <td>32</td>\n",
       "      <td>37</td>\n",
       "      <td>2</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>23e1ac07-8810-4141-8adf-7aa4579104f4</td>\n",
       "      <td>325</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>4</td>\n",
       "      <td>40;99;217;182</td>\n",
       "      <td>100.0;100.0;100.0;95.83</td>\n",
       "      <td>4</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>37</td>\n",
       "      <td>41</td>\n",
       "      <td>3</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              read_name  read_seq_length  enzyme_matches  \\\n",
       "0  6ff9aeed-b85d-415a-aa59-3ff86bbe727d              532               1   \n",
       "1  5faf5728-3f2e-4d37-8f62-574673a1662e              286               0   \n",
       "2  4e3a9167-dac1-429e-9351-ce55485a4fae              416               0   \n",
       "3  f536f761-c128-4cde-a247-8b3a924b3399              261               0   \n",
       "4  23e1ac07-8810-4141-8adf-7aa4579104f4              325               0   \n",
       "\n",
       "  enzyme_pos  total_bc_matches     forward_bc_pos  \\\n",
       "0        492                 8  34;93;152;211;270   \n",
       "1         -1                 4                 35   \n",
       "2         -1                 5         92;151;210   \n",
       "3         -1                 3              38;96   \n",
       "4         -1                 4      40;99;217;182   \n",
       "\n",
       "                forward_bc_score  forward_bc_macthes reverse_bc_pos  \\\n",
       "0  100.0;100.0;100.0;100.0;100.0                   5    343;402;458   \n",
       "1                          100.0                   1    108;167;226   \n",
       "2              100.0;100.0;100.0                   3        283;342   \n",
       "3                    100.0;97.87                   2            183   \n",
       "4        100.0;100.0;100.0;95.83                   4             -1   \n",
       "\n",
       "    reverse_bc_score  reverse_bc_macthes  mean_base_quality  \\\n",
       "0  100.0;100.0;100.0                   3                 30   \n",
       "1  100.0;100.0;100.0                   3                 38   \n",
       "2        100.0;100.0                   2                 36   \n",
       "3              95.83                   1                 32   \n",
       "4                 -1                   0                 37   \n",
       "\n",
       "   median_base_quality  min_base_quality  max_base_quality  \n",
       "0                   35                 2                50  \n",
       "1                   42                 2                50  \n",
       "2                   41                 4                50  \n",
       "3                   37                 2                50  \n",
       "4                   41                 3                50  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def find_restriction_sites(read_seq, rb):\n",
    "    \"\"\"A function to find all restriction sites \n",
    "    in a sequence \"\"\"\n",
    "    search_results = rb.search(Seq(read_seq))\n",
    "    sites = list(search_results.values())[0]\n",
    "    n_sites = len(sites)\n",
    "    if n_sites == 0:\n",
    "        sites = [-1]\n",
    "\n",
    "    site_str = \";\".join([str(x) for x in sites])\n",
    "    return site_str, n_sites\n",
    "\n",
    "\n",
    "def find_fuzzy(seq, tgt, min_sim):\n",
    "  \"\"\"\n",
    "  Finds all fuzzy matches of the target sequence within the main sequence recursively.\n",
    "\n",
    "  Args:\n",
    "      seq: The main sequence to search.\n",
    "      tgt: The target sequence to find fuzzy matches for.\n",
    "      min_sim: Minimum similarity threshold (0-100).\n",
    "\n",
    "  Returns:\n",
    "      A list of tuples containing the matching character, matched substring, start and end positions, and similarity score.\n",
    "  \"\"\"\n",
    "  matches = []\n",
    "  best_match = None\n",
    "\n",
    "  # Find the best alignment using a sliding window and update best_match\n",
    "  for i in range(len(seq) - len(tgt) + 1):\n",
    "    sub = seq[i:i + len(tgt)]\n",
    "    score = fuzz.partial_ratio(tgt, sub)\n",
    "    if score >= min_sim and (not best_match or score > best_match[4]):\n",
    "      best_match = (seq[i], sub, i, i + len(tgt) - 1, score)\n",
    "\n",
    "  # If a match is found, record it and continue recursively\n",
    "  if best_match:\n",
    "    matches.append(best_match)\n",
    "    rem_seq = seq[:best_match[2]] + seq[best_match[3] + 1:]\n",
    "    matches.extend(find_fuzzy(rem_seq, tgt, min_sim))\n",
    "\n",
    "  return matches\n",
    "\n",
    "\n",
    "def find_barcodes(read_seq, barcode, min_sim=92):\n",
    "    \"\"\"A function to find barcode sequences in reads \"\"\"\n",
    "    barcode_len = len(barcode)\n",
    "\n",
    "    matches = find_fuzzy(read_seq, barcode, min_sim)\n",
    "    n_matches = len(matches)\n",
    "    \n",
    "    if len(matches) == 0:\n",
    "        pos = [-1]\n",
    "        scores = [-1]\n",
    "    else:\n",
    "        pos = []\n",
    "        scores = []\n",
    "        for i, match in enumerate(matches):\n",
    "            _, _, start, _, score = match\n",
    "            # correct alignment positions for striped barcodes\n",
    "            offset = barcode_len * i            \n",
    "            start = start + offset\n",
    "            pos.append(start)\n",
    "            scores.append(score)\n",
    "\n",
    "    pos_str = \";\".join([str(x) for x in pos])\n",
    "    score_str = \";\".join([str(round(x, 2)) for x in scores])\n",
    "    return pos_str, score_str, n_matches\n",
    "    \n",
    "\n",
    "def get_sequence_report(fpath, rb, barcode, barcode_rc):\n",
    "    \"\"\"A function to return read-level information \n",
    "    from fastq files \"\"\"\n",
    "    report_df = []\n",
    "    \n",
    "    count = -1\n",
    "    stop = 1000\n",
    "    for read in pysam.FastxFile(fpath):\n",
    "        count += 1\n",
    "        if count == stop:\n",
    "            break\n",
    "\n",
    "        # get read information\n",
    "        read_seq = read.sequence\n",
    "        read_seq_length = len(read_seq)\n",
    "        quals = read.get_quality_array()\n",
    "\n",
    "        # find the restrictin enzyme sites\n",
    "        enzyme_pos, enzyme_matches = find_restriction_sites(read_seq, rb)\n",
    "\n",
    "        # find forward barcodes\n",
    "        fw_pos_str, fw_score_str, fw_n_matches = find_barcodes(read_seq, barcode)\n",
    "\n",
    "        # find reverse complement barcodes\n",
    "        rv_pos_str, rv_score_str, rv_n_matches = find_barcodes(read_seq, barcode_rc)\n",
    "\n",
    "        # total BC\n",
    "        total_bc_matches = fw_n_matches + rv_n_matches\n",
    "\n",
    "        row = {\n",
    "            'read_name' : read.name,\n",
    "            'read_seq_length' : read_seq_length,\n",
    "            'enzyme_matches' : enzyme_matches,\n",
    "            'enzyme_pos' : enzyme_pos,\n",
    "            'total_bc_matches' : total_bc_matches,\n",
    "            'forward_bc_pos' : fw_pos_str,\n",
    "            'forward_bc_score' : fw_score_str,\n",
    "            'forward_bc_macthes' : fw_n_matches,\n",
    "            'reverse_bc_pos' : rv_pos_str,\n",
    "            'reverse_bc_score' : rv_score_str,\n",
    "            'reverse_bc_macthes' : rv_n_matches,\n",
    "            'mean_base_quality' : int(np.mean(quals)),\n",
    "            'median_base_quality' : int(np.median(quals)),\n",
    "            'min_base_quality' : np.min(quals),\n",
    "            'max_base_quality' : np.max(quals),\n",
    "        }\n",
    "        report_df.append(row)\n",
    "    return pd.DataFrame(report_df)\n",
    "    \n",
    "\n",
    "fastq_path = \"/scratch/indikar_root/indikar1/cstansbu/scpc_test/fastq/o3b01.raw.fastq\"\n",
    "barcode_path = \"../config/barcodes.txt\"\n",
    "enzyme = 'NlaIII'\n",
    "\n",
    "# set up restriction enzyme\n",
    "rb = Restriction.RestrictionBatch([enzyme])\n",
    "\n",
    "# set up get barcodes\n",
    "barcode_id = os.path.basename(fastq_path).split(\".\")[0][2:]\n",
    "code_df = pd.read_csv(barcode_path)\n",
    "barcode = code_df[code_df['cell_id'] == barcode_id]['barcode'].values[0]\n",
    "barcode_rc = str(Seq(barcode).reverse_complement())\n",
    "\n",
    "df = get_sequence_report(fastq_path, rb, barcode, barcode_rc)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c6b22b0a-210e-442a-9651-25ddd784b394",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_name</th>\n",
       "      <th>read_seq_length</th>\n",
       "      <th>enzyme_matches</th>\n",
       "      <th>enzyme_pos</th>\n",
       "      <th>total_bc_matches</th>\n",
       "      <th>forward_bc_pos</th>\n",
       "      <th>forward_bc_score</th>\n",
       "      <th>forward_bc_macthes</th>\n",
       "      <th>reverse_bc_pos</th>\n",
       "      <th>reverse_bc_score</th>\n",
       "      <th>reverse_bc_macthes</th>\n",
       "      <th>mean_base_quality</th>\n",
       "      <th>median_base_quality</th>\n",
       "      <th>min_base_quality</th>\n",
       "      <th>max_base_quality</th>\n",
       "      <th>barcode_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6ff9aeed-b85d-415a-aa59-3ff86bbe727d</td>\n",
       "      <td>532</td>\n",
       "      <td>1</td>\n",
       "      <td>492</td>\n",
       "      <td>8</td>\n",
       "      <td>34;93;152;211;270</td>\n",
       "      <td>100.0;100.0;100.0;100.0;100.0</td>\n",
       "      <td>5</td>\n",
       "      <td>343;402;458</td>\n",
       "      <td>100.0;100.0;100.0</td>\n",
       "      <td>3</td>\n",
       "      <td>30</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>50</td>\n",
       "      <td>b01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5faf5728-3f2e-4d37-8f62-574673a1662e</td>\n",
       "      <td>286</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>4</td>\n",
       "      <td>35</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1</td>\n",
       "      <td>108;167;226</td>\n",
       "      <td>100.0;100.0;100.0</td>\n",
       "      <td>3</td>\n",
       "      <td>38</td>\n",
       "      <td>42</td>\n",
       "      <td>2</td>\n",
       "      <td>50</td>\n",
       "      <td>b01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4e3a9167-dac1-429e-9351-ce55485a4fae</td>\n",
       "      <td>416</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>5</td>\n",
       "      <td>92;151;210</td>\n",
       "      <td>100.0;100.0;100.0</td>\n",
       "      <td>3</td>\n",
       "      <td>283;342</td>\n",
       "      <td>100.0;100.0</td>\n",
       "      <td>2</td>\n",
       "      <td>36</td>\n",
       "      <td>41</td>\n",
       "      <td>4</td>\n",
       "      <td>50</td>\n",
       "      <td>b01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>f536f761-c128-4cde-a247-8b3a924b3399</td>\n",
       "      <td>261</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>3</td>\n",
       "      <td>38;96</td>\n",
       "      <td>100.0;97.87</td>\n",
       "      <td>2</td>\n",
       "      <td>183</td>\n",
       "      <td>95.83</td>\n",
       "      <td>1</td>\n",
       "      <td>32</td>\n",
       "      <td>37</td>\n",
       "      <td>2</td>\n",
       "      <td>50</td>\n",
       "      <td>b01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>23e1ac07-8810-4141-8adf-7aa4579104f4</td>\n",
       "      <td>325</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>4</td>\n",
       "      <td>40;99;217;182</td>\n",
       "      <td>100.0;100.0;100.0;95.83</td>\n",
       "      <td>4</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>37</td>\n",
       "      <td>41</td>\n",
       "      <td>3</td>\n",
       "      <td>50</td>\n",
       "      <td>b01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              read_name  read_seq_length  enzyme_matches  \\\n",
       "0  6ff9aeed-b85d-415a-aa59-3ff86bbe727d              532               1   \n",
       "1  5faf5728-3f2e-4d37-8f62-574673a1662e              286               0   \n",
       "2  4e3a9167-dac1-429e-9351-ce55485a4fae              416               0   \n",
       "3  f536f761-c128-4cde-a247-8b3a924b3399              261               0   \n",
       "4  23e1ac07-8810-4141-8adf-7aa4579104f4              325               0   \n",
       "\n",
       "  enzyme_pos  total_bc_matches     forward_bc_pos  \\\n",
       "0        492                 8  34;93;152;211;270   \n",
       "1         -1                 4                 35   \n",
       "2         -1                 5         92;151;210   \n",
       "3         -1                 3              38;96   \n",
       "4         -1                 4      40;99;217;182   \n",
       "\n",
       "                forward_bc_score  forward_bc_macthes reverse_bc_pos  \\\n",
       "0  100.0;100.0;100.0;100.0;100.0                   5    343;402;458   \n",
       "1                          100.0                   1    108;167;226   \n",
       "2              100.0;100.0;100.0                   3        283;342   \n",
       "3                    100.0;97.87                   2            183   \n",
       "4        100.0;100.0;100.0;95.83                   4             -1   \n",
       "\n",
       "    reverse_bc_score  reverse_bc_macthes  mean_base_quality  \\\n",
       "0  100.0;100.0;100.0                   3                 30   \n",
       "1  100.0;100.0;100.0                   3                 38   \n",
       "2        100.0;100.0                   2                 36   \n",
       "3              95.83                   1                 32   \n",
       "4                 -1                   0                 37   \n",
       "\n",
       "   median_base_quality  min_base_quality  max_base_quality barcode_id  \n",
       "0                   35                 2                50        b01  \n",
       "1                   42                 2                50        b01  \n",
       "2                   41                 4                50        b01  \n",
       "3                   37                 2                50        b01  \n",
       "4                   41                 3                50        b01  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paths = [\n",
    "     \"/scratch/indikar_root/indikar1/cstansbu/scpc_test/fastq/o3b01.raw.fastq\",\n",
    "     \"/scratch/indikar_root/indikar1/cstansbu/scpc_test/fastq/o1b04.raw.fastq\",\n",
    "]\n",
    "barcode_path = \"../config/barcodes.txt\"\n",
    "enzyme = 'NlaIII'\n",
    "\n",
    "# set up restriction enzyme\n",
    "rb = Restriction.RestrictionBatch([enzyme])\n",
    "\n",
    "df = []\n",
    "for fastq_path in paths:\n",
    "    # set up get barcodes\n",
    "    barcode_id = os.path.basename(fastq_path).split(\".\")[0][2:]\n",
    "    code_df = pd.read_csv(barcode_path)\n",
    "    barcode = code_df[code_df['cell_id'] == barcode_id]['barcode'].values[0]\n",
    "    barcode_rc = str(Seq(barcode).reverse_complement())\n",
    "\n",
    "    tmp = get_sequence_report(fastq_path, rb, barcode, barcode_rc)\n",
    "    tmp['barcode_id'] = barcode_id\n",
    "    df.append(tmp)\n",
    "\n",
    "df = pd.concat(df)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cd3c62f-fbf4-435c-aff5-e4d2d7098cf6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e56fd60e-15e4-4db6-8715-f6835bb11a21",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "da38b063-76d7-4c20-869e-748b269c3951",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "'break' outside loop (668683560.py, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn[5], line 1\u001b[0;36m\u001b[0m\n\u001b[0;31m    break\u001b[0m\n\u001b[0m    ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m 'break' outside loop\n"
     ]
    }
   ],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca35e575-33e8-4cbb-b101-44e022425c3f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "079d888c-67fd-4e95-93f4-2d6752153584",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e21fa2a2-159d-45b8-b24c-ee3f5bbb42dd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fdf45431-c318-4c40-ba8a-68c0b9009e3e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5ff40a0-5d98-4372-a691-661c0e631ccc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_first_position(position_rec):\n",
    "    if \";\" in position_rec:\n",
    "        return int(position_rec.split(\";\")[0])\n",
    "    else:\n",
    "        return int(position_rec)\n",
    "\n",
    "def get_last_position(position_rec):\n",
    "    if \";\" in position_rec:\n",
    "        return int(position_rec.split(\";\")[-1])\n",
    "    else:\n",
    "        return int(position_rec)\n",
    "\n",
    "def print_report(df):\n",
    "    \"\"\"A function to summarize the sequence report \"\"\"\n",
    "    n = len(df)\n",
    "    print(f\"\\n###### SUMMARY SECTION ######\")\n",
    "    print(f\"Total sequences: {len(df)}\")\n",
    "\n",
    "    print(f\"\\n###### SEQUENCE LENGTH ######\")\n",
    "    print(f\"Mean sequence length: {int(df['read_seq_length'].mean())} bp\")\n",
    "    print(f\"Median sequence length: {int(df['read_seq_length'].median())} bp\")\n",
    "    print(f\"Min sequence length: {int(df['read_seq_length'].min())} bp\")\n",
    "    print(f\"Max sequence length: {int(df['read_seq_length'].max())} bp\")\n",
    "\n",
    "    print(f\"\\n###### BASE QUALITY ######\")\n",
    "    print(f\"Mean base quality: {int(df['mean_base_quality'].mean())}\")\n",
    "    print(f\"Median base quality: {int(df['mean_base_quality'].median())}\")\n",
    "    print(f\"Min base quality: {int(df['mean_base_quality'].min())}\")\n",
    "    print(f\"Max base quality: {int(df['mean_base_quality'].max())}\")\n",
    "\n",
    "    print(f\"\\n###### ENZYME SECTION ######\")\n",
    "    n_enzymes = pd.DataFrame(df['enzyme_matches'].value_counts())\n",
    "    for idx, row in n_enzymes.iterrows():\n",
    "        count = row['count']\n",
    "        print(f\"Sequences with {idx} NlaIII sites: {count} ({100*(count / n):.2f}%)\")\n",
    "\n",
    "    print(f\"\\n###### FORWARD BARCODE SECTION ######\")\n",
    "    # get the first barcode position\n",
    "    masked_df = df[df['forward_bc_pos'] != '-1']\n",
    "    first_position = masked_df['forward_bc_pos'].apply(lambda x: get_first_position(x)).values\n",
    "    last_position = masked_df['forward_bc_pos'].apply(lambda x: get_last_position(x)).values\n",
    "    print(f\"Average first barcode position on read: {int(np.mean(first_position))}\")\n",
    "    print(f\"Average last barcode position on read: {int(np.mean(last_position))}\")\n",
    "    print()\n",
    "    \n",
    "    n_bc = pd.DataFrame(df['forward_bc_macthes'].value_counts())\n",
    "    for idx, row in n_bc.iterrows():\n",
    "        count = row['count']\n",
    "        print(f\"Sequences with {idx} barcodes: {count} ({100*(count / n):.2f}%)\")\n",
    "\n",
    "\n",
    "    print(f\"\\n###### REVERSE BARCODE SECTION ######\")\n",
    "    n_bc = pd.DataFrame(df['reverse_bc_macthes'].value_counts())\n",
    "    for idx, row in n_bc.iterrows():\n",
    "        count = row['count']\n",
    "        print(f\"Sequences with {idx} barcodes: {count} ({100*(count / n):.2f}%)\")\n",
    "    \n",
    "    \n",
    "    # print(df['enzyme_matches'].value_counts())\n",
    "\n",
    "\n",
    "print_report(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23f4f5cf-674f-466a-a137-1c29e83812c2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
